### PR TITLE
C++: Fix use-after-cast bug in SimpleRangeAnalysis

### PR DESCRIPTION
**Pull request by jbj**
This should fix the FP reported at https://discuss.lgtm.com/t/c-false-positive-comparison-result-is-always-the-same/1857.

### CODE REVIEW

**Comment by jbj** [comment]
It turns out there was a difference in `PointlessComparison.expected`. The new alert looks correct to me, but I'm not sure if it's correct for the right reasons. Maybe a CPP-Differences job would be in order.

File: cpp/ql/test/library-tests/rangeanalysis/SimpleRangeAnalysis/test.cpp [hunk 1]
```diff
@@ -68,3 +68,13 @@ bool test3(bool b, int x, int y) {
 
   return b || (bool)t;
 }
+
+void use_after_cast(unsigned char c)
+{
+  unsigned short c_times_2 = c + c;
+  if ((unsigned char)c_times_2 == 0)
+  {
+    c_times_2; // BUG: upper bound should be 510, not 0
```
Comment (hmakholm) [inline cpp/ql/test/library-tests/rangeanalysis/SimpleRangeAnalysis/test.cpp]: A better upper bound would be 256, but that is probably too much to ask :-)

File: cpp/ql/test/library-tests/rangeanalysis/SimpleRangeAnalysis/upperBound.expected [hunk 2]
```diff
@@ -459,3 +459,8 @@
 | test.cpp:67:13:67:14 | xb | 1.0 |
 | test.cpp:69:10:69:10 | b | 1.0 |
 | test.cpp:69:21:69:21 | t | 3.0 |
+| test.cpp:74:30:74:30 | c | 255.0 |
+| test.cpp:74:34:74:34 | c | 255.0 |
+| test.cpp:75:22:75:30 | c_times_2 | 510.0 |
+| test.cpp:77:5:77:13 | c_times_2 | 0.0 |
+| test.cpp:79:3:79:11 | c_times_2 | 255.0 |
```
Comment (hmakholm) [inline cpp/ql/test/library-tests/rangeanalysis/SimpleRangeAnalysis/upperBound.expected]: This is not the correct result!

**Comment by jbj** [comment]
The failing test shows that we're losing good results in some cases. This is probably because we're not modelling that it's fine to cast to wider types. I'll see if that's easily fixed.

**Comment by jbj** [comment]
I've fixed the test failure in a way that should also fix other `Conversion` issues, and I've started a CPP-Differences job at https://jenkins.internal.semmle.com/job/Query-Changes/job/CPP-Differences/534/

**Comment by jbj** [comment]
CPP-Differences job showed some unexpected changes that I'm still trying to understand.

**Comment by jbj** [comment]
Most of the unexpected CPP-Differences changes were caused by some missing calls to `getUnspecifiedType` in my new code. I've fixed that and started [a new CPP-Differences job](https://jenkins.internal.semmle.com/job/Query-Changes/job/CPP-Differences/541/).

**Comment by jbj** [comment]
[The CPP-Differences results](https://jenkins.internal.semmle.com/job/Query-Changes/job/CPP-Differences/541/) are in:

```
=== Overall changes (new / fixed) ===

                                                                           20 / 6


=== Changes per project (new / fixed) ===

linux                                                                      19 / 5
jdk8u                                                                       0 / 1
git                                                                         1 / 0


=== Changes per query (new / fixed) ===

PointlessComparison.ql                                                     11 / 4
SnprintfOverflow.ql                                                         9 / 0
ComparisonWithCancelingSubExpr.ql                                           0 / 1
StackAddressEscapes.ql                                                      0 / 1
```

The four "fixed" alerts in `PointlessComparison.ql` all look like instances of the use-after-cast bug that this PR is intended to fix. The "new" alerts from that query seem to arise because we now consider some enums to be `>= 0`. I don't know if it's the case for all enums, but at least it gives these 11 new alerts on defensive code in the Linux kernel. In any case, I think the "fixed" alerts are more obviously wrong than the "new" alerts.

There are some new results for `SnprintfOverflow.ql` in Linux. To me, they look similar to other code that tends to be flagged by this query.

@kevinbackhouse and @geoffw0, what do you think about these changes?

**Comment by jbj** [comment]
I just checked, and both Clang and GCC (in C mode) use `unsigned int` as the underlying type for enums unless there's a constant with negative value, in which case they use `int`. That means the new alerts about enums are technically correct although they rely on implementation-defined behaviour.

**Comment by geoffw0** [comment]
> The CPP-Differences results are in:

The diff results look good to me except:
 - as you noted, there are some new results about range checks on enums that I'm not convinced about.
 - linux / drivers / gpu / drm / i915 / intel_dp.c line 1481 'Potentially overflowing call to snprintf' looks like a FP to me (I don't think this is high priority stuff, certainly shouldn't block this PR).
 
> That means the new alerts about enums are technically correct although they rely on implementation-defined behaviour.

What will the situation be on LGTM?  If my code is written exclusively for a platform that uses `signed int` for enums, I'm not going to be interested in these alerts just because LGTM perhaps happens to use `unsigned int`.

**Comment by jbj** [comment]
I don't think we can justify those enum alerts based on platform-specific choices of signed vs. unsigned. Do you think we can justify them by saying that if `x` has type `enum t`, then the value of `x` can be assumed to be one of the constants in `enum t`? I'm not too confident that we can justify the alerts like that. The alerts I've seen are just flagging defensive code, not pointing to anything that's wrong.

If you agree, then I'll find some way to make the alerts go away. They appeared because the library changed, but I think the library is fine. We could exclude them syntactically in the query just by excluding all results that involve an enum access.

**Comment by jbj** [comment]
I think I've restored the previous behaviour for enums now. It turns out there was already an exclusion rule for them, but I had broken it.

**Comment by geoffw0** [comment]
Then LGTM.

I'm not confident enough to merge this without a second opinion (or a new run of the differences job) - because this isn't an area of code I know well.

**Comment by jbj** [comment]
I'll start a CPP-Differences job

**Comment by jbj** [comment]
There's a CPP-Differences job running here: https://jenkins.internal.semmle.com/job/Query-Changes/job/CPP-Differences/543

**Comment by geoffw0** [comment]
Diff LGTM.

**Comment by geoffw0** [comment]
Two test failures though:
```
[2019-04-08T11:23:53.535Z]     ql/cpp/ql/test/library-tests/rangeanalysis/SimpleRangeAnalysis/lowerBound.ql: Test failure (Expected output does not match)
[2019-04-08T11:23:53.535Z]     ql/cpp/ql/test/library-tests/rangeanalysis/SimpleRangeAnalysis/upperBound.ql: Test failure (Expected output does not match)
```

**Comment by jbj** [comment]
The changes on this branch have affected `SnprintfOverflow.ql` in ways I don't yet understand. I'll try to get to the bottom of it.

**Comment by jbj** [comment]
The central question in this PR is this: when we learn the result of a comparison like `e1 >= e2`, what do we learn about `e1`? If `x` is a 16-bit `short`, then we don't learn anything useful about `x` from `(char)x >= 2`. We do learn something useful from `(int)x >= 2`. The `SnprintfOverflow` query started acting up because I changed the library to say we don't learn anything useful from `(unsigned int)x >= (unsigned int)2` either. Essentially we learn that either `x` is either negative or at least 2, but that's not helpful for the range analysis. With my latest change, I've restored the old behaviour of saying we learn that `x` is at least 2. That's technically wrong, but it's at least no more wrong than current `master`.

**Comment by jbj** [comment]
I'm also running https://jenkins.internal.semmle.com/job/Query-Changes/job/CPP-Differences/547 to make sure there are no more unforeseen consequences.

**Comment by jbj** [comment]
I've pushed a small commit, hopefully the last one, to fix a comment in a test that I forgot to update.

The CPP-Differences job produced 0 changes, which is surprising to me, but the new lines in `upperBound.expected` still look right, so I assume this PR still fixes the original problem. I think I got [the parameters](https://jenkins.internal.semmle.com/job/Query-Changes/job/CPP-Differences/547/parameters/) right in the CPP-Differences job. [Here's](https://git.semmle.com/jonas/code/compare/78a8a0e89dd28d1a24354c5436dc15092944e42d...8abf689b029f329d863b0b85f6f6d4e892da4939) the diff between the two SHAs from those parameters.

**Pull request merged**